### PR TITLE
Partial Neighbor Allreduce Implementation under NCCL

### PR DESCRIPTION
**Pull request by ybc1991**

### CODE REVIEW

File: bluefog/common/mpi_controller.cc [hunk 1]
```diff
@@ -263,40 +263,44 @@ void MPIController::NeighborAllgather(TensorTableEntry& entry) {
...
+      send_check_buf[size + recv_rank] = true;
+    int ret_code = MPI_Allgather(send_check_buf, size * 2, MPI_C_BOOL,
+                                 recv_check_buf, size * 2, MPI_C_BOOL, comm);
     if (ret_code != MPI_SUCCESS) {
-        throw std::runtime_error(
-            "MPI_Allgather (for dynamic neighbor_allreduce negotiation) failed, see MPI output "
-            "for details.");
+      throw std::runtime_error(
+          "MPI_Allgather (for dynamic neighbor_allreduce negotiation) failed, "
+          "see MPI output "
```
[the remark this answers is not in the document]
Comment (ybc1991) [reply bluefog/common/mpi_controller.cc:287]: That is auto formate. I think it exceeds 80 chars.

File: test/torch_ops_test.py [hunk 2]
```diff
@@ -298,6 +298,7 @@ def test_neighbor_allreduce_avg_precision(self):
                 (reduced_tensor.data - sum_value).abs().max() == 0
             ), "bf.neighbor_allreduce (avg) produces incorrect reduced tensor"
 
+    @unittest.skip("Still failed with error.")
```
[the remark this answers is not in the document]
Comment (ybc1991) [reply test/torch_ops_test.py:301]: ok

File: bluefog/common/mpi_controller.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (hanbinhu) [reply bluefog/common/mpi_controller.cc:287]: But, why is there a line break between this line and the next one?

File: bluefog/common/nccl_controller.cc [hunk 3]
```diff
@@ -470,9 +471,162 @@ void NCCLController::NeighborAllgather(TensorTableEntry& entry) {
...
+
+  void* buffer_data = (void*)entry.output->data();
+
+  // We need to explicitly set the device here.
+  with_device device_guard(entry.device);
+
+  // If only partial sending is enabled, the following code block checks whether
+  // the sending and recieving neighbors match each other when enable_topo_check
+  // is set to be True.
+  bool is_topo_check_fail = CheckNeighborSendRecvPattern(
```
[the remark this answers is not in the document]
Comment (ybc1991) [reply bluefog/common/nccl_controller.cc:509]: I don't think so. These values are not on GPU.

File: bluefog/common/nccl_controller.cc [hunk 4]
```diff
@@ -470,9 +471,162 @@ void NCCLController::NeighborAllgather(TensorTableEntry& entry) {
...
+    }
+    for (int send_rank : *entry.send_neighbors) {
+      NCCLCHECK(ncclSend(sendbuf, num_elements, GetNCCLDataType(entry.tensor),
+                         send_rank, nccl_ctx_.nccl_comm, nccl_ctx_.stream));
+    }
+  }
+  ncclGroupEnd();
+
+  auto tid = std::this_thread::get_id();
+  // TODO: use thread pool or single thread for callbacks
```
[the remark this answers is not in the document]
Comment (ybc1991) [reply bluefog/common/nccl_controller.cc:549]: yes.

File: bluefog/common/nccl_controller.cc [hunk 5]
```diff
@@ -470,9 +471,162 @@ void NCCLController::NeighborAllgather(TensorTableEntry& entry) {
...
+  if (is_topo_check_fail) {
+    entry.callback(Status::InvalidArgument(
+        "Send and recv neighbors dont' match in neighbor "
+        "allreduce with partial send/recv request."));
+  }
+
+  timeline_ptr->ActivityStart(entry.tensor_name, "COMMUNICATE");
+#if NCCL_MINOR > 6
+  ncclGroupStart();
+  if (entry.send_neighbors->empty()) {
```
[the remark this answers is not in the document]
Comment (ybc1991) [reply bluefog/common/nccl_controller.cc:521]: That is partially true. There are 6 effective lines and 4 lines are different. The gain is not huge.
But from the readability view, I prefer to keep the current style.

File: bluefog/common/mpi_controller.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (ybc1991) [reply bluefog/common/mpi_controller.cc]: You are right. Done

**Pull request merged**